### ISSUE TITLE
Automatic & manual version sync / continous deployment

### ISSUE THREAD

**Issue by rducom** [DataDog/dd-trace-dotnet#674]
Hi,

I'm currently evaluating the Data dog APM for our production, and I noted the following in the 1.12 release : 

> NuGet package Datadog.Trace is still used for manual instrumentation or to configure the Tracer in code. **Please keep the NuGet package version in sync with the installed Tracer** if you will be mixing automatic and manual instrumentation in the same application.

This restriction is really blocking for our use case : we deploy our apps by copying files on VM with IIS, and we are on a continuous deployment strategy (about 10 to 20 deploys per day).
- We have a .Net framework legacy monolith on one side, witch is deployed on 10 different VM per cluster. In this case we need to synchronize app deployment with infrastructure deployment, which seems a bad idea.
- We have a lot of services and apps running on .NetCore on the other side, witch are deployed on some shared VM (about 12 apps / services per VM). In this case, we need to deploy all the apps at the same time, with the same Datadog.Trace nuget version, and at the same time synchronize the datadog-dotnet-apm-XXX-x64.msi deployment... Honestly this is impossible.

If we have to maintain the version correlation between datadog-dotnet-apm-XXX-x64.msi and the Datadog.Trace nuget package, we just can't upgrade either the nuget or the automatic instrumentation while continuously deploying. 

This restriction seems really blocking for us. 

One solution would be to release both packages (automatic & manual) on independent lifecycles (like dotnet/aspnet do with the .Abstraction packages). But on your side, you should have a proper "assembly binding" strategy, or retro-compat strategy, in order to support upgrades on one side at a time.

I didn't tested the behavior while upgrading the automatic or manual instrumentation on one side only, but since you made the warning, I would like to know if you have planned the mentioned continuous delivery scenario, and what are the current available workaround ?

Thanks in advance,

**Comment by rducom** [comment]
One missing piece in my explanation : we are currently using the Datadog.Trace nuget in parallel of the automatic instrumentation for 2 important things : 
- log / apm correlation since we have a custom log implementation, we need to call `CorrelationIdentifier.SpanId` / `TraceId` in order to inject the info in our logs
- and the ability to force Span selection with `Span?.SetTag(Tags.ManualKeep, "true");` in some specific cases (when we see some specific errors)

And we plan to use custom Spans in our code later, but that's another story.

**Comment by lucaspimentel** [comment]
@rducom: Thanks for your feedback! We are aware that this is an issue for users who use both the msi and the nuget packages at the same time. We don't have a firm timeline yet, but we are [already working](https://github.com/DataDog/dd-trace-dotnet/pull/635) on lifting this restriction in most cases.

The idea is similar to the dotnet Abstraction libraries you mentioned. We'll have core libraries that will change rarely and can keep a pretty constant assembly version, and those can be shared between both the automatic and manual instrumentation libraries.

**Comment by rducom** [comment]
Hello @lucaspimentel ! Since https://github.com/DataDog/dd-trace-dotnet/pull/635 have been closed, what are the plans on this issue ?

**Comment by macrogreg** [comment]
We are currently working on support for Open Telemetry and .NET Activities. Once shipped, you will be able to use those standard .NET APIs to create custom spans without the versioning conflict described here. The work is ongoing, but we do not have a specific release date yet.

**Comment by macrogreg** [comment]
Internal tracker:
https://datadoghq.atlassian.net/browse/DOTNET-21

**Comment by pierotibou** [comment]
Our [last release](https://github.com/DataDog/dd-trace-dotnet/releases/tag/v2.0.1) now allows the msi and nuget to be out of sync. Regarding this fix:

- on .Net Core, you need to upgrade custom instrumentation first. The fix will work with automatic instrumentation starting v1.28.8

- on .Net Framework, you need both the custom and automatic tracer to run on v2.x for the conflict to be solved. Futures upgrades won't show this issue anymore though.
